nbclient/client.py

```diff
@@ -425,6 +424,33 @@ def setup_kernel(self, **kwargs):
         finally:
             self._cleanup_kernel()
 
+    @contextmanager
+    def graceful_shutdown(self):
```

Comment:
This approach is a bit ham-fisted, but it does seem to work. So in cases of the friendly signals, cleanup should happen. Anything else is unfortunately impossible to handle, and I can't imagine the old way of cleanup did that consistently (or at all).

What's annoying is that without the signal handler, the finally clause of `async_setup_kernel` doesn't get called, since apparently the interrupt goes straight to the ioloop and cancels all the tasks on it without raising exceptions inside functions (I'm learning more than I wanted to about asyncio...lol). But when the handler is used, it causes a separate failure in `async_execute`, which causes that finally clause to run, running cleanup twice.

I'm only using `atexit` here since Windows doesn't do signals (may have to tweak this if the signal handler function doesn't even _exist_).

Reply:
Yeah I agree that the double cleanup should be fairly harmless as the cleanup functions become no-ops.

Thanks so much for digging deep into this. It is appreciated checking these conditions thoroughly!

Reply:
Not exactly a no-op, since the shutdown message is still sent, but that's not really worth worrying about IMO.